### PR TITLE
[CS2113-W12-1] ModMate

### PR DESCRIPTION
**Pull request by vegetablestabber**
The app aims to provide extra functionality for a given NUSMods timetable, such as checking common empty time slots among two timetables, calculating workload hours, assigning travel time and calculating when the user should leave home to reach class on time etc.

### CODE REVIEW

File: src/main/java/modmate/CommandCenter.java [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        BreakPeriod breakPeriod = new BreakPeriod(label, new Period(day, start, end));
+        boolean success = currentUser.addBreakToTimetable(timetableName, breakPeriod);
+
+        if (success) {
+            System.out.println("Break [" + label + "] added to timetable '" + timetableName + "'.");
+        } else {
+            System.out.println("Timetable '" + timetableName + "' not found.");
+        }
+    }
+}
```
Comment (Alfred-Goh02) [inline src/main/java/modmate/CommandCenter.java]: Consider separating parsing functions into different files from the command.

File: src/main/java/modmate/CommandCenter.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
+package modmate;
+
```
Comment (Alfred-Goh02) [inline src/main/java/modmate/CommandCenter.java]: Very nice usage of the API and data processing, well done!

File: src/main/java/modmate/Main.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+public class Main {
+
+    /**
+     * The main command loop of the application that processes user input
+     * and executes corresponding actions based on the command received.
+     *
+     * @param args Command-line arguments (for logging configuration).
+     */
+    public static void main(String[] args) {
+
```
Comment (Alfred-Goh02) [inline src/main/java/modmate/Main.java:23]: Overall very good usage of packages to separate classes for different feature

File: src/main/java/modmate/log/Log.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                        isNewSession = false;
+                    }
+                    out.println(message);
+                }
+            } catch (IOException e) {
+                System.err.println("Failed to write log to file: " + e.getMessage());
+            }
+        }
+    }
+
```
Comment (Alfred-Goh02) [inline src/main/java/modmate/log/Log.java]: Consider using Java in built logging library, or you can add more levels to your logging such as WARN, INFO etc.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,38 +1,246 @@
...
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
 
+## **Design**
 
-## Product scope
-### Target user profile
+### **Architecture**
 
-{Describe the target user profile}
+<img src="images/ArchitectureDiagram.png" width="280" />
```
Comment (ojassurana) [inline docs/DeveloperGuide.md]: The image cannot be viewed. Perhaps make sure there is an image at the directory here?

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,38 +1,246 @@
...
+      a formatted String is returned which shows, the timetable name, all added modules and module details. If timetable is not found, 
+      an error message, "Timetable 'X' not found", will be displayed.
+
+- **Design Rationale**: Usage of List<Schedule> allowed for maintainable insertion order of timetables and straightforward iteration. 
+Moreover, with the timetable duplication checking it allows for unnecessary object creation and potential naming conflicts.
+- **Possible Extensions**: Merging of multiple timetables
+
+The following UML Sequence diagram effectively shows how these two methods are implemented with the inclusion of user inputs of `createtimetable` and `timetable` 
+for `addTimetable(String timetableName)` and `getTimetable(String timetableName)` respectively.
+![image.png](image/image.png)
```
Comment (ojassurana) [inline docs/DeveloperGuide.md]: The image cannot be seen. Perhaps ensure that the directory of the image is appropriate?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,38 +1,246 @@
...
+      2. `getTimetable(String timetableName)` - Method to display the user's timetables (mods included in the timetable)
+     These methods are supported by `Schedule` class for storage structure
+      
+      `addTimetable(String timetableName)` first checks whether there are any duplicate timetables with the same name. It checks through a list called `List<Schedule> timetables`
+      and if duplicate timetable found, an error message, "Timetable 'X' already exists", will be displayed. If no duplicate timetable found, a new Schedule object is instantiated  
+      and a new timetable of `timetableName` will be created. It is added to `timetables` list, afterwhich a success message, "Timetable 'X' created successfully", is displayed.
+
+      `getTimetable(String timetableName)` searches through a list called `timetables`, as mentioned earlier. If timetable with corresponding `timetableName` is found,  
+      a formatted String is returned which shows, the timetable name, all added modules and module details. If timetable is not found, 
+      an error message, "Timetable 'X' not found", will be displayed.
```
Comment (ojassurana) [inline docs/DeveloperGuide.md]: The formatting for the explanations of how each function works is some what cluttered. Perhaps, cleaning up the formatting—such as consistent indentation and correcting minor typos (e.g., “afterwhich” to “after which”)—will improve readability.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,38 +1,246 @@
...
+- It attempts to retrieve the module using modFromCodeOrName.
+- If a match is found, it displays the detailed module information and logs the action.
+- If no match is found, it informs the user and logs the failure.
+
+This feature involves the following key operations:
+- `viewMod(inputCodeOrName)`: Main entry point for viewing module details.
+- `modFromCodeOrName(modCodeOrNameGiven)`: Attempts to find a module based on the provided code or name.
+- `fetchModuleByCode(moduleCode)`: Fetches module details from the NUSMods API.
+
+#### Example Usage Scenario
```
Comment (ojassurana) [inline docs/DeveloperGuide.md:93]: This section is quiet brief. Maybe adding sample command inputs and the expected outputs (or even a screenshot of a terminal session) could help clarify how the feature works?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,38 +1,246 @@
...
+- **Possible Extensions**: Merging of multiple timetables
+
+The following UML Sequence diagram effectively shows how these two methods are implemented with the inclusion of user inputs of `createtimetable` and `timetable` 
+for `addTimetable(String timetableName)` and `getTimetable(String timetableName)` respectively.
+![image.png](image/image.png)
+
+#### **API Integration**
+- **Purpose**: Fetches module data from the NUSMods API.
+- **Implementation**: The `NUSModsAPI` class interacts with the API using `HttpUtil` for network requests and `JsonParser` for processing responses.
+- **Rationale**: Separating API logic into a dedicated class ensures that external dependencies are isolated from core application logic.
```
Comment (ojassurana) [inline docs/DeveloperGuide.md]: In the scenario there's a network outage and NUSModsAPI is not reachable, how will the software handle this exception? If the software doesn't work it might potentially be counted as a major bug. Perhaps maybe shift to offline data storage?

**Review by ojassurana — COMMENTED** [review]
Mostly LGTM, just minor changes such as lack of images.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,38 +1,308 @@
...
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
 
+## **Design**
 
-## Product scope
-### Target user profile
+### **Architecture**
 
-{Describe the target user profile}
+<img src="images/ArchitectureDiagram.png" width="280" />
```
Comment (HuaZhenting) [inline docs/DeveloperGuide.md:54]: Is this architecture diagram properly inserted? I cannot view it in your DG.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,38 +1,308 @@
...
+
+### **Advanced Scenarios**
+1. **Concurrent Modifications**:
+    - Simulate concurrent user actions (e.g., adding and removing modules in quick succession) and verify that the application handles them correctly.
+2. **Stress Testing**:
+    - Continuously add and remove modules in rapid succession to test the application's stability.
+3. **Cross-Component Interactions**:
+    - Test commands that involve multiple components (e.g., adding a module and immediately bookmarking it) to ensure seamless integration.
+
+These instructions are tailored to the application's structure and functionality, ensuring comprehensive manual testing.
```
Comment (HuaZhenting) [inline docs/DeveloperGuide.md]: I appreciate that your DG is very neat and well-structured!

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,38 +1,308 @@
...
+#### **Purpose**
+
+Allows users to create, update, and view multiple academic schedules.
+
+#### **Implementation**
+
+Two core methods in the `User` class:
+1. `addTimetable(String timetableName)` - Method to add a new timetable for the user
+2. `getTimetable(String timetableName)` - Method to display the user's timetables (mods included in the timetable)
+
```
Comment (HuaZhenting) [inline docs/DeveloperGuide.md:160]: Would it be better if you could provide the expected input format or give brief examples here? It would help make the explanation clearer.

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,38 +1,308 @@
...
+### **Target User Profile**
+
+* NUS students who need a tool to manage their academic modules and timetables.
+* Users comfortable with Command Line Interfaces (CLI).
+
+### **Value Proposition**
+
+* Provides a lightweight and efficient way for NUS students to plan and manage their academic schedules and module information, tailored to the unique requirements of the NUS modular system.
+
```
Comment (HuaZhenting) [inline docs/DeveloperGuide.md:299]: I like your user profile and value proposition! It's clear and reasonable.

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (itsjoelha) [reply docs/DeveloperGuide.md:54]: I can view the diagram in the DG.

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,38 +1,322 @@
...
+These instructions are tailored to the application's structure and functionality, ensuring comprehensive manual testing:
+
+1. **Launching the Application**:
+    - Run the `Main` class located in `src/main/java/modmate/Main.java`.
+    - Verify that the application starts without errors and displays the welcome message.
+
+2. **Adding Modules**:
+    - Enter a valid `add` command (e.g., `add CS2113`) and verify that the module is added to the in-memory timetable.
+    - Test edge cases such as adding duplicate modules or invalid module codes (e.g., `add XYZ123`) and confirm appropriate error messages are displayed.
```
Comment (itsjoelha) [inline docs/DeveloperGuide.md]: Perhaps you can include the expected outputs.

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,38 +1,322 @@
...
-{Describe the value proposition: what problem does it solve?}
+![image.png](img/jahnavi/image.png)
+
+### Search Module Feature
+
+#### Proposed Implementation
+{: .no_toc }
+
+![Search Mod Class Diagram](img/harii/searchmod_class.svg)
+
```
Comment (itsjoelha) [inline docs/DeveloperGuide.md:236]: I like the explanation of the process of how it works.

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,38 +1,322 @@
...
+    - Verify that the bookmarked module is displayed when using the `getbookmarks` command.
+    - Test edge cases such as bookmarking non-existent modules.
+
+7. **Timetable Creation**:
+    - Test the `createtimetable` command to generate a new timetable.
+    - Verify that the timetable is initialized correctly and does not overwrite existing data unless explicitly confirmed.
+
+### Error Handling
+
+1. Enter invalid commands (e.g., `add123`) and confirm that the application displays appropriate error messages.
```
Comment (itsjoelha) [inline docs/DeveloperGuide.md:360]: Perhaps you can include an example error message.

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,38 +1,322 @@
...
+- It attempts to retrieve the module using modFromCodeOrName.
+- If a match is found, it displays the detailed module information and logs the action.
+- If no match is found, it informs the user and logs the failure.
+
+This feature involves the following key operations:
+- `viewMod(inputCodeOrName)`: Main entry point for viewing module details.
+- `modFromCodeOrName(modCodeOrNameGiven)`: Attempts to find a module based on the provided code or name.
+- `fetchModuleByCode(moduleCode)`: Fetches module details from the NUSMods API.
+
+#### Example Usage Scenario
```
Comment (itsjoelha) [inline docs/DeveloperGuide.md:93]: I like the logical step by step process.

File: docs/img/harii/searchmod_class.svg
Comment (tejaskumar0) [inline docs/img/harii/searchmod_class.svg:1]: This diagram is comprehensive but you could consider adding more descriptive labels to the arrows between classes.

File: docs/team/img/image.png
Comment (tejaskumar0) [inline docs/team/img/image.png:1]: Great job covering both addModToTimetable and removeModFromTimetable flows in the same sequence diagram-it gives a complete picture of user interactions.

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,38 +1,382 @@
...
+3b. If no module is found, an error message is displayed, and the failure is logged.
+
+![img.png](img/luke/img3.png)
+#### Design Considerations
+{: .no_toc }
+
+1. Data Retrieval Method
+    - Alternative 1 (Current Choice): Fetch from External API Each Time
+        - Pros: Ensures the latest module details are always retrieved.
+        - Cons: API calls may introduce latency.
```
Comment (Basudeb2005) [inline docs/DeveloperGuide.md:111]: Could this be expanded to include specific performance targets or benchmarks that the application aims to meet?

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,38 +1,382 @@
...
+| `*`      | advanced user    | export my timetable to a file    | share it with others or back it up        |
 
 ## Non-Functional Requirements
 
-{Give non-functional requirements}
+1. The application should work on any mainstream OS with Java 17 installed.
+2. The application should log all user actions and errors to a file for debugging purposes.
 
-## Glossary
+## Appendix: Instructions for Manual Testing
```
Comment (Basudeb2005) [inline docs/DeveloperGuide.md:321]: The testing instructions don't specify any particular environment requirements. Are there specific Java versions, operating systems, or other environment factors that affect test results?

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1,38 +1,382 @@
...
+
+## Design
+
+### Architecture
+
+![Architecture Diagram](img/architecture.svg)
+
+The ***Architecture Diagram*** above provides a high-level overview of the system. It shows the main components of the application and how they interact with each other.
+
+**Main components of the architecture:**
```
Comment (Basudeb2005) [inline docs/DeveloperGuide.md:57]: It appears that the listing of main components might benefit from the inclusion of NUSModsAPI, User, JsonParser, HttpUtil, and Schedule.

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -1,38 +1,382 @@
...
+{:toc}
 
 ## Acknowledgements
 
-{list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the original source as well}
+* The application integrates with the [NUSMods API](https://nusmods.com/) to fetch module data. Special thanks to the NUSMods team for providing their publicly available API.
+* Command parsing logic is inspired by [AddressBook-Level3](https://github.com/se-edu/addressbook-level3).
+* UML diagrams were created using [PlantUML](https://plantuml.com/).
+
+## Glossary
```
Comment (Basudeb2005) [inline docs/DeveloperGuide.md:25]: Enhance the glossary section to include all technical terms and acronyms used in the document. This is particularly helpful for new developers or contributors who may not be familiar with the domain-specific language.

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -1,38 +1,386 @@
...
+- Usage of `List<Schedule>` allowed for maintainable insertion order of timetables and straightforward iteration.
+- Moreover, with the timetable duplication checking it allows for unnecessary object creation and potential naming conflicts.
+
+#### Possible Extensions: Merging of multiple timetables
+{: .no_toc }
+
+- The following UML Sequence diagram effectively shows how these two methods are implemented with the inclusion of user inputs of `createtimetable` and `timetable` for `addTimetable(String timetableName)` and `getTimetable(String timetableName)` respectively.
+
+![image.png](img/jahnavi/image.png)
+
```
Comment (HalFentise) [inline docs/DeveloperGuide.md:170]: I think you can split two method in two diagrams

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -1,38 +1,386 @@
...
+This feature involves the following key operations:
+- `execute(String[] args, User currentUser)`: Main entry point for executing the search command.
+- `getSearchResults(String searchTerm)`: Searches for modules matching the query and retrieves their details.
+
+#### Sequence Diagram
+{: .no_toc }
+
+![Search Mod Sequence Diagram](img/harii/searchmod_seq.svg)
+
```
Comment (HalFentise) [inline docs/DeveloperGuide.md:253]: there are no time bar for CommandLine

File: docs/DeveloperGuide.md [hunk 24]
```diff
@@ -1,38 +1,386 @@
...
-|--------|----------|---------------|------------------|
-|v1.0|new user|see usage instructions|refer to them when I forget how to use the application|
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+| Priority | As a …​          | I want to …​                     | So that I can …​                          |
+| -------- | ---------------- | -------------------------------- | ----------------------------------------- |
+| `* * *`  | new user         | see usage instructions           | refer to them when I forget how to use the application |
+| `* * *`  | student          | search for a module by name or code | quickly find information about a specific module |
+| `* *`    | student          | add a module to my timetable     | keep track of my academic schedule        |
+| `* *`    | student          | view all modules in my timetable | get an overview of my academic workload   |
+| `*`      | advanced user    | export my timetable to a file    | share it with others or back it up        |
```
Comment (HalFentise) [inline docs/DeveloperGuide.md:314]: i like you use priority instead of version, its more clear

File: docs/DeveloperGuide.md [hunk 25]
```diff
@@ -1,38 +1,386 @@
...
+- It attempts to retrieve the module using modFromCodeOrName.
+- If a match is found, it displays the detailed module information and logs the action.
+- If no match is found, it informs the user and logs the failure.
+
+This feature involves the following key operations:
+- `viewMod(inputCodeOrName)`: Main entry point for viewing module details.
+- `modFromCodeOrName(modCodeOrNameGiven)`: Attempts to find a module based on the provided code or name.
+- `fetchModuleByCode(moduleCode)`: Fetches module details from the NUSMods API.
+
+#### Example Usage Scenario
```
Comment (HalFentise) [inline docs/DeveloperGuide.md:93]: for this ####title i think you can change it to other format because now it is now showing very clearly.
